packages/parse-mapping-lookup/src/parser.ts

```diff
@@ -0,0 +1,139 @@
+import debugModule from "debug";
+const debug = debugModule("parse-mapping-lookup:parser");
+
+import {
+  regexp,
+  anyCharOf,
+  noCharOf,
+  stringLen,
+  string,
+  int,
+  float
+} from "parjs";
+import {
+  qthen,
+  map,
+  many,
+  stringify,
+  between,
+  or,
+  then
+} from "parjs/combinators";
+
+import {
+  access,
+  expression,
+  identifier,
+  literal,
+  lookup,
+  pointer
+} from "./ast";
+
+/*
+ * Identifier
+ */
+
+const identifierP = regexp(/[a-zA-Z][a-zA-Z0-9]*/).pipe(
+  map(([name]) => identifier({ name }))
+);
+
+/*
+ * Literal
+ */
+
+namespace Strings {
+  // borrowed from parjs JSON example
+  // https://github.com/GregRos/parjs/blob/master/src/examples/json.ts
+
+  const escapeChars = {
+    '"': `"`,
+    "\\": "\\",
+    "/": "/",
+    "f": "\f",
+    "n": "\n",
+    "r": "\r",
+    "t": "\t"
+  };
+
+  const escapeCharP = anyCharOf(Object.keys(escapeChars).join()).pipe(
+    map(char => escapeChars[char] as string)
+  );
+
+  // A unicode escape sequence is "u" followed by exactly 4 hex digits
+  const unicodeEscapeP = string("u").pipe(
+    qthen(
+      stringLen(4).pipe(
+        map(str => parseInt(str, 16)),
+        map(x => String.fromCharCode(x))
+      )
+    )
+  );
+
+  // Any escape sequence begins with a \
+  const escapeP = string("\\").pipe(
+    qthen(escapeCharP.pipe(or(unicodeEscapeP)))
+  );
+
+  // Here we process regular characters vs escape sequences
+  const stringEntriesP = escapeP.pipe(or(noCharOf('"')));
+
+  // Repeat the char/escape to get a sequence, and then put between quotes to get a string
+  export const stringP = stringEntriesP.pipe(many(), stringify(), between('"'));
+}
+
+const numberP = int().pipe(or(float()));
+
+const stringP = Strings.stringP;
+
+const literalP = numberP.pipe(
```

Comment:
Solidity has other sorts of literals that I think need to be accounted for.  Specifically:
1. Boolean literals (`true` and `false`)
2. Hex string literals -- I assume static bytestring literals `0x1234` (and address literals) fall under numeric literals, but there's also dynamic bytestring / hex string literals `hex"deadbeef"`.
3. It also might be worth supporting Solidity units here, as those are effectively part of numeric literals?

Reply:
Oops, one more I forgot about -- string literals are normally restricted to ASCII, but you can also do unicode string literals as `unicode"..."`; better account for that too.

Reply:
Oops, one more I forgot about -- string literals are normally restricted to ASCII, but you can also do unicode string literals as `unicode"..."`; better account for that too.

Reply:
- [ ] Boolean literals
- [ ] `hex"deadbeef"` literals
- [ ] `unicode"..."` literals
- [ ] Solidity units (` gwei`, etc.)

Reply:
Hm so unicode literals are a bit annoying, it seems, because Parjs seems to just assume its input is in the BMP (See their README section [Parsing Unicode](https://github.com/GregRos/parjs#parsing-unicode)).

It should be doable to restrict `""` literals to ASCII, but it might be a bit annoying to write the parser for that?

What do you think of this:
- Exclude `unicode"..."` as unsupported for now (since BMP restriction suggests that we shouldn't make the promise of full-unicode support)
- Don't restrict `""` to ASCII in the meantime, to afford _some way_ of specifying literals that include BMP characters.

Alternatively, if you want to be a sticker about `""` meaning ASCII, we could just define our own literal `bmp""` to be more explicit.

Thoughts?

Reply:
I don't think we need to actually enforce the ASCII restriction.  Especially as that's relatively recent, older Solidity doesn't make a distinction.

I think it's also OK to exclude `unicode"..."` for now.  Proper Unicode handling will be a pain regardless of how we do it, so maybe it's OK if we just don't support it all properly for now.  (Especially the `\xNN` escape codes...)

Reply:
> (Especially the \xNN escape codes...)

I don't want to know

Reply:
Too bad I'm going to explain anyway! :P  In Solidity, you can do "\xNN" to insert a particular byte.  Note, not a particular *character* -- a particular *byte*, regardless of whether or not this makes valid UTF-8.  However, if it's invalid UTF-8, then you won't be able to use it as a `string`, only as a `bytes` or `bytesN` (since string literals can be used as either).  So, e.g., `"unicode\xA1"` is not the same as `unicode"\u00A1"` or `unicode"¡"`, but rather is an error if used as a `string` rather than a `bytes`/`bytesN`.

Reply:
Ha! So does Solidity handle this and error if you try to assign such a literal to a `string`?

> Too bad I'm going to explain anyway! 

Good cause I was lying.

Reply:
The Solidity compiler detects it upfront.  It *is* possible to put bad UTF-8 inside a string at runtime; there are no runtime checks on whether your UTF-8 is good.  But attempting to do so directly from a literal (whether `unicode` or `hex` or whatever) will trip the compile-time check and cause a compile error.  If you really want to put bad UTF-8 inside a `string`, you'll need to be sneakier about it.  (Putting it in a `bytes` and the converting that `bytes` to a `string` will do the trick. :) )